### PR TITLE
Accelerate augmentations

### PR DESCRIPTION
**Pull request by FGeri**
### Description
This PR speeds up augmentation moduls

Original:
color_jitter : 14.622583630176651 (kornia_time:57.04273818107322,torchvision_time:3.9010026971809566)
hflip : 3.476565444566317 (kornia_time:0.26084241503849626,torchvision_time:0.07502876594662666)
vflip : 13.391427033125542 (kornia_time:0.4837029566988349,torchvision_time:0.036120344419032335)

Accelerated:
color_jitter : 1.396787808236991 (kornia_time:5.459571750368923,torchvision_time:3.908662230707705)
hflip : 1.8372702968448464 (kornia_time:0.1345923920162022,torchvision_time:0.07325671799480915)
vflip : 0.5296634456533842 (kornia_time:0.018525363877415657,torchvision_time:0.03497572662308812)

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Review by shijianjian — APPROVED** [review]
Interesting to see ```torch.flip``` is slower. Can you share your testing code in this PR? Will try it out then merge.

**Comment by FGeri** [comment]
> Interesting to see `torch.flip` is slower. Can you share your testing code in this PR? Will try it out then merge.
Here is a  github issue about it: https://github.com/pytorch/pytorch/issues/16424

Here is the benchmark code, nothing fancy:
`import torch
import kornia.augmentation as augmentation
import torchvision.transforms as transforms
import numpy as np
from PIL import Image
import timeit

n = 100
NUM_OF_IMGS = 1

kornia_augmentations = {
"center_crop": augmentation.CenterCrop((256,256),resample=0),
"color_jitter": augmentation.ColorJitter(brightness=0.1, contrast=0.1, saturation=0.1, hue=0.1,p=1.),
"grayscale": augmentation.RandomGrayscale(p=1.0),
"affine": augmentation.RandomAffine(5., translate=(0.1,0.2), scale=(1.0,1.5), shear=(-5.,5.),p=1.0,resample=0),
"hflip": augmentation.RandomHorizontalFlip(p=1.0),
"vflip": augmentation.RandomVerticalFlip(p=1.0),
"random_crop": augmentation.RandomCrop((256,256),resample=0,p=1.0),
"resized_crop": augmentation.RandomResizedCrop((256,256), scale=(0.08, 1.0), ratio=(0.75, 1.3333333333333333),p=1.0),
# "erase": augmentation.RandomErasing(scale = (0.02, 0.33), ratio = (0.3, 3.3), p=1.0),
"random_perspective": augmentation.RandomPerspective(distortion_scale=0.5, p=1.0),
"random_rot": augmentation.RandomRotation(15., resample=0,p=1.0)
}

tensor = torch.ones((NUM_OF_IMGS,3,720,1280)).float()
def kornia(key):
    out = kornia_augmentations[key](tensor)

trochvision_augmentations = {
    "center_crop": transforms.CenterCrop((256,256)),
    "color_jitter": transforms.ColorJitter(brightness=0.1, contrast=0.1, saturation=0.1, hue=0.1),
    "grayscale": transforms.Grayscale(num_output_channels=3),
    "affine": transforms.RandomAffine(5., translate=(0.1,0.2), scale=(1.0,1.5), shear=(-5.,5.),),
    "hflip": transforms.RandomHorizontalFlip(p=1.0),
    "vflip": transforms.RandomVerticalFlip(p=1.0),
    "random_crop": transforms.RandomCrop((256,256)),
    "resized_crop": transforms.RandomResizedCrop((256,256), scale=(0.08, 1.0), ratio=(0.75, 1.3333333333333333)),
    # "erase": transforms.RandomErasing(scale = (0.02, 0.33), ratio = (0.3, 3.3), p=1.0),
    "random_perspective": transforms.RandomPerspective(distortion_scale=0.5, p=1.0, interpolation=2, fill=0),
    "random_rot": transforms.RandomRotation(15.)
}

pil_img = Image.fromarray(np.ones((720,1280,3),dtype=np.uint8))
def torchvision(key):  
    for _ in range(NUM_OF_IMGS):
        out = trochvision_augmentations[key](pil_img)

if __name__ == '__main__':
    print(f"Based on {n} measurements")
    for key in kornia_augmentations.keys():
        kornia_time = timeit.timeit(f"kornia('{key}')", setup="from __main__ import kornia",number=n)
        torchvision_time = timeit.timeit(f"torchvision('{key}')", setup="from __main__ import torchvision",number=n)
        print(f"{key} : {kornia_time/torchvision_time} (kornia_time:{kornia_time},torchvision_time:{torchvision_time})")`

**Comment by shijianjian** [comment]
Brilliant. On my MAC CPU, It accelerated 10x for hflip and vflip and about 4x for colorjitter.

File: kornia/color/hsv.py [hunk 1]
```diff
@@ -65,14 +65,14 @@ def hsv_to_rgb(image: torch.Tensor) -> torch.Tensor:
...
-    out[out == 4] = torch.stack((t, p, v), dim=-3)[out == 4]
-    out[out == 5] = torch.stack((v, p, q), dim=-3)[out == 5]
+    hi = hi.long()
+    indices: torch.Tensor = torch.stack([hi, hi + 6, hi + 12], dim=-3)
+    out = torch.stack((
+        v, q, p, p, t, v,
+        t, v, v, q, p, p,
+        p, p, t, v, v, q,
+    ), dim=-3)
+    out = torch.gather(out, -3, indices)
```
Comment (edgarriba) [inline kornia/color/hsv.py:75]: is there any way to index directly from a (v,q,p,t) tensor instead of stacking repeated patterns ?

File: kornia/geometry/transform/flips.py [hunk 2]
```diff
@@ -123,8 +123,8 @@ def hflip(input: torch.Tensor) -> torch.Tensor:
         torch.Tensor: The horizontally flipped image tensor
 
     """
-
-    return torch.flip(input, [-1])
+    w = input.shape[-1]
+    return input[..., torch.arange(w - 1, -1, -1)]
```
Comment (edgarriba) [inline kornia/geometry/transform/flips.py:127]: have you tested in GPU ? this I guess will crash, 
```python
w = input.shape[-1]
index = torch.arange(w - 1, -1, -1, device=input.device, dtype=input.dtype)
return input[..., index]
```

Comment (shijianjian) [reply kornia/geometry/transform/flips.py:127]: Yes. It makes more sense.

@FGeri Can you do another PR for a quick fix? I do not have a GPU machine by hand.

File: kornia/geometry/transform/flips.py [hunk 3]
```diff
@@ -139,4 +139,5 @@ def vflip(input: torch.Tensor) -> torch.Tensor:
 
     """
 
-    return torch.flip(input, [-2])
+    h = input.shape[-2]
+    return input[..., torch.arange(h - 1, -1, -1), :]
```
Comment (vadimkantorov) [inline kornia/geometry/transform/flips.py:143]: Will this still work if input is on GPU and torch.arange output is CPU?

Comment (edgarriba) [reply kornia/geometry/transform/flips.py:143]: this was fixed https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/flips.py#L149

Did you encounter any issues around this ?

Comment (vadimkantorov) [reply kornia/geometry/transform/flips.py:143]: Ah, okay. Then it should be fine!
